### PR TITLE
Laboratory work #4, Anna Belyavskaya - 23FPL2

### PR DESCRIPTION
**Pull request by AnnaBeliavskaya**

### CODE REVIEW

**Comment by EIzharikov** [comment]
Nothing to review, I hope at least the memes will be funny

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -26,6 +27,9 @@ def get_paragraphs(text: str) -> list[str]:
     Returns:
         list[str]: Paragraphs from document.
     """
+    if not text:
+        raise ValueError
+    return text.split('\n')
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:32]: there can be blank strings after this operation, so result after parsing this input
```
hello, my friend!\n
\n
hello!
```
will be ['hello, my friend!','hello!']

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -40,6 +44,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        self._avg_doc_len = -1.0
+        super().__init__(self._corpus)
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:49]: call super init at start of `__init__`

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -67,6 +79,10 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        vector = self._calculate_bm25(tokenized_document)
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:82]: check input arg

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -81,6 +97,14 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list):
+            raise ValueError
+        bm25vector = [0.0] * len(self._vocabulary)
+        bm25 = calculate_bm25(self._vocabulary, tokenized_document, self._idf_values,\
+                              1.5, 0.75, self._avg_doc_len, len(tokenized_document))
+        for index, word in enumerate(self._vocabulary):
+            bm25vector[index] = bm25[word]
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:106]: use `self_token2ind`

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -113,6 +141,24 @@ def put_corpus(self, corpus: Corpus) -> None:
...
+        for text in corpus:
+            tokenized_text = self._tokenizer.tokenize(text)
+            if tokenized_text:
+                tokenized_texts.append(tokenized_text)
+                self.__documents.append(text)
+        if not tokenized_texts:
+            raise ValueError
+        self._vectorizer.set_tokenized_corpus(tokenized_texts)
+        self._vectorizer.build()
+        try:
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:157]: for what reason ?

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -182,6 +246,14 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        tokenized_query = self._db.get_tokenizer().tokenize(query)
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:249]: check input arguments are correct

**Review by EIzharikov — COMMENTED** [review]
Do not be slow! Go go! Fix CI and all my comments. Good luck!

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py:49]: this way it doesn't work

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py:249]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py:32]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py:82]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py:106]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -170,6 +232,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
+        super().__init__(self._db.get_vectorizer(), self._db.get_tokenizer())
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: call super init firstly

File: lab_4_retrieval_w_clustering/settings.json [hunk 8]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 6
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/settings.json:2]: i can not see 8 here, where is it?

**Review by EIzharikov — COMMENTED** [review]
everything is fine, but where is 8? I know that you can reach it! Good luck!

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (AnnaBeliavskaya) [reply lab_4_retrieval_w_clustering/main.py]: done

**Comment by EIzharikov** [comment]
So, there is almost nothig new to review, you fixed init, nice! Be ready for Friday! Good luck!

File: lab_4_retrieval_w_clustering/settings.json (hunk 8 above)
Comment (EIzharikov) [reply lab_4_retrieval_w_clustering/settings.json:2]: go away

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -26,6 +27,9 @@ def get_paragraphs(text: str) -> list[str]:
     Returns:
         list[str]: Paragraphs from document.
     """
+    if not text:
+        raise ValueError
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: write error message

**Pull request closed without merging**